test/jdk/java/net/httpclient/HttpResponseLimitingTest.java

```diff
@@ -0,0 +1,396 @@
+        }
+
+    }
+
+    @AfterAll
+    static void closeServerClientPairs() {
+        for (var pair : new ServerClientPair[]{HTTP1, HTTPS1, HTTP2, HTTPS2}) {
+            pair.client.close();
+            pair.server.stop();
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("sufficientCapacities")
+    void testSuccessOnSufficientCapacityForByteArray(ServerClientPair pair, long sufficientCapacity) throws Exception {
+        HttpResponse<byte[]> response = pair.request(BodyHandlers.ofByteArray(), sufficientCapacity);
+        verifyHeaders(response.headers());
+        assertArrayEquals(RESPONSE_BODY, response.body());
+    }
+
+    @ParameterizedTest
+    @MethodSource("sufficientCapacities")
+    void testSuccessOnSufficientCapacityForInputStream(ServerClientPair pair, long sufficientCapacity) throws Exception {
+        HttpResponse<InputStream> response = pair.request(BodyHandlers.ofInputStream(), sufficientCapacity);
+        verifyHeaders(response.headers());
+        try (InputStream responseBodyStream = response.body()) {
+            byte[] responseBodyBuffer = new byte[RESPONSE_BODY.length];
+            int responseBodyBufferLength = responseBodyStream.read(responseBodyBuffer);
+            assertEquals(responseBodyBuffer.length, responseBodyBufferLength);
+            assertArrayEquals(RESPONSE_BODY, responseBodyBuffer);
+        }
+    }
+
+    static Arguments[] sufficientCapacities() {
+        long minExtremeCapacity = RESPONSE_BODY.length;
+        long maxExtremeCapacity = Long.MAX_VALUE;
+        long nonExtremeCapacity = RANDOM.nextLong(minExtremeCapacity + 1, maxExtremeCapacity);
+        return capacityArgs(minExtremeCapacity, nonExtremeCapacity, maxExtremeCapacity);
+    }
+
+    @ParameterizedTest
+    @MethodSource("insufficientCapacities")
+    void testFailureOnInsufficientCapacityForByteArray(ServerClientPair pair, long insufficientCapacity) {
+        BodyHandler<byte[]> handler = responseInfo -> {
+            verifyHeaders(responseInfo.headers());
+            return BodySubscribers.limiting(BodySubscribers.ofByteArray(), insufficientCapacity);
+        };
+        var exception = assertThrows(IOException.class, () -> pair.request(handler, insufficientCapacity));
+        assertEquals(exception.getMessage(), "body exceeds capacity: " + insufficientCapacity);
+    }
+
+    @ParameterizedTest
+    @MethodSource("insufficientCapacities")
+    void testFailureOnInsufficientCapacityForInputStream(ServerClientPair pair, long insufficientCapacity) throws Exception {
+        HttpResponse<InputStream> response = pair.request(BodyHandlers.ofInputStream(), insufficientCapacity);
+        verifyHeaders(response.headers());
+        try (InputStream responseBodyStream = response.body()) {
+            var exception = assertThrows(
+                    IOException.class,
+                    () -> responseBodyStream.read(new byte[RESPONSE_BODY.length]));
```

Comment:
Here too, I would suggest reading till EOF and expecting IOException at some point while doing so, for the same reason I noted in a review comment above.

Reply:
Implemented in 21ae1079aaa3730317ef1bb11277e929dbd6174d.